drivers/input/touchscreen/himax.c

```diff
@@ -0,0 +1,462 @@
+		{
+			.addr = client->addr,
+			.flags = 0,
+			.len = 1,
+			.buf = &command,
+		},
+		{
+			.addr = client->addr,
+			.flags = I2C_M_RD,
+			.len = length,
+			.buf = data,
+		}
+	};
+	// clang-format on
+
+	ret = i2c_transfer(client->adapter, msg, ARRAY_SIZE(msg));
+
+	if (ret != ARRAY_SIZE(msg)) {
+		dev_err(&ts->client->dev, "i2c read failed: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int himax_i2c_write(struct himax_ts_data *ts, u8 command, u8 *data,
+			   u8 length)
+{
+	int ret;
+	u8 buf[16];
+
+	struct i2c_msg msg = {
+		.addr = ts->client->addr,
+		.flags = 0,
+		.len = length + 1,
+		.buf = buf,
+	};
+
+	if (length + 1 > sizeof(buf)) {
+		dev_err(&ts->client->dev, "i2c write too large: %uB\n", length);
+		return -ENOMEM;
+	}
+
+	buf[0] = command;
+	memcpy(buf + 1, data, length);
+
+	ret = i2c_transfer(ts->client->adapter, &msg, 1);
+
+	if (ret != 1) {
+		dev_err(&ts->client->dev, "i2c write failed: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int himax_i2c_write_u8(struct himax_ts_data *ts, u8 command, u8 data)
+{
+	return himax_i2c_write(ts, command, &data, sizeof(data));
+}
+
+static int himax_register_read(struct himax_ts_data *ts, u32 address, u32 *dst)
+{
+	int ret = 0;
+
+	__le32 address_data = cpu_to_le32(address);
+	__le32 dst_data;
+
+	ret = himax_i2c_write(ts, HIMAX_CMD_REG_SET_ADDR, (u8 *)&address_data,
+			      sizeof(address_data));
+	if (ret)
+		return ret;
+
+	ret = himax_i2c_write_u8(ts, HIMAX_CMD_REG_INIT_READ, 0x00);
+	if (ret)
+		return ret;
+
+	ret = himax_i2c_read(ts, HIMAX_CMD_REG_READ_VALUE, (u8 *)&dst_data,
+			     sizeof(dst_data));
+	if (ret)
+		return ret;
+
+	*dst = le32_to_cpu(dst_data);
+	return 0;
+}
+
+static int himax_read_input_event(struct himax_ts_data *ts,
+				  struct himax_event *event)
+{
+	return himax_i2c_read(ts, HIMAX_CMD_READ_EVENT, (u8 *)event,
+			      sizeof(*event));
+}
+
+static void himax_reset(struct himax_ts_data *ts)
+{
+	gpiod_set_value(ts->gpiod_rst, 0);
+	msleep(20);
+	gpiod_set_value(ts->gpiod_rst, 1);
+}
+
+static int himax_read_product_id(struct himax_ts_data *ts, u32 *product_id)
+{
+	int error = himax_register_read(ts, HIMAX_REG_PRODUCT_ID, product_id);
+
+	if (error)
+		return error;
+
+	*product_id >>= 8;
+	return 0;
+}
+
+static int himax_check_product_id(struct himax_ts_data *ts)
+{
+	int error;
+	u32 product_id;
+
+	error = himax_read_product_id(ts, &product_id);
+	if (error)
+		return error;
+
+	switch (product_id) {
+	case HIMAX_ID_83112A:
+	case HIMAX_ID_83112B:
+		return 0;
+
+	default:
+		dev_err(&ts->client->dev, "Unknown product id: %x\n",
+			product_id);
+		return -ENODEV;
+	}
+}
+
+static int himax_setup_gpio(struct himax_ts_data *ts)
+{
+	int error;
+
+	ts->gpiod_rst = devm_gpiod_get(&ts->client->dev, HIMAX_GPIO_RST_NAME,
+				       GPIOD_OUT_LOW);
+	if (IS_ERR(ts->gpiod_rst)) {
+		error = PTR_ERR(ts->gpiod_rst);
+		dev_err(&ts->client->dev, "Failed to get %s GPIO: %d\n",
+			HIMAX_GPIO_RST_NAME, error);
+		return error;
```

Comment:
Use something like this instead of the 3 lines
```
		return dev_err_probe(dev, PTR_ERR(haptics->reset_gpio),
				     "Failed to get reset gpio\n");
```